lib/sectorbuilder/sectorbuilder.go

```diff
@@ -68,31 +163,164 @@ func (sb *SectorBuilder) Destroy() {
 	sectorbuilder.DestroySectorBuilder(sb.handle)
 }
 
-func (sb *SectorBuilder) AddPiece(pieceKey string, pieceSize uint64, file io.Reader) (uint64, error) {
-	f, werr, err := toReadableFile(file, int64(pieceSize))
+func (sb *SectorBuilder) AcquireSectorId() (uint64, error) {
+	sb.idLk.Lock()
+	defer sb.idLk.Unlock()
+
+	id, err := sectorbuilder.AcquireSectorId(sb.handle)
 	if err != nil {
 		return 0, err
 	}
-
-	sectorID, err := sectorbuilder.AddPieceFromFile(sb.handle, pieceKey, pieceSize, f)
+	err = sb.ds.Put(lastSectorIdKey, []byte(fmt.Sprint(id)))
 	if err != nil {
 		return 0, err
 	}
+	return id, nil
+}
+
+func (sb *SectorBuilder) AddPiece(pieceSize uint64, sectorId uint64, file io.Reader, existingPieceSizes []uint64) (PublicPieceInfo, error) {
+	f, werr, err := toReadableFile(file, int64(pieceSize))
+	if err != nil {
+		return PublicPieceInfo{}, err
+	}
+
+	ret := sb.rlimit()
+	defer ret()
+
+	stagedFile, err := sb.stagedSectorFile(sectorId)
+	if err != nil {
+		return PublicPieceInfo{}, err
+	}
+
+	_, _, commP, err := sectorbuilder.StandaloneWriteWithAlignment(f, pieceSize, stagedFile, existingPieceSizes)
+	if err != nil {
+		return PublicPieceInfo{}, err
+	}
+	/*if writeUnpadded != pieceSize {
+		return PublicPieceInfo{}, xerrors.Errorf("writeUnpadded != pieceSize: %d != %d", writeUnpadded, pieceSize)
+	}*/
+
+	if err := stagedFile.Close(); err != nil {
+		return PublicPieceInfo{}, err
+	}
+
+	if err := f.Close(); err != nil {
+		return PublicPieceInfo{}, err
+	}
 
-	return sectorID, werr()
+	return PublicPieceInfo{
+		Size:  pieceSize,
+		CommP: commP,
+	}, werr()
 }
 
 // TODO: should *really really* return an io.ReadCloser
 func (sb *SectorBuilder) ReadPieceFromSealedSector(pieceKey string) ([]byte, error) {
+	ret := sb.rlimit()
+	defer ret()
+
 	return sectorbuilder.ReadPieceFromSealedSector(sb.handle, pieceKey)
 }
 
-func (sb *SectorBuilder) SealSector(sectorID uint64, ticket SealTicket) (SealedSectorMetadata, error) {
-	return sectorbuilder.SealSector(sb.handle, sectorID, ticket)
+func (sb *SectorBuilder) SealPreCommit(sectorID uint64, ticket SealTicket, pieces []PublicPieceInfo) (RawSealPreCommitOutput, error) {
+	ret := sb.rlimit()
+	defer ret()
+
+	cacheDir, err := sb.sectorCacheDir(sectorID)
+	if err != nil {
+		return RawSealPreCommitOutput{}, err
+	}
+
+	sealedPath, err := sb.sealedSectorPath(sectorID)
+	if err != nil {
+		return RawSealPreCommitOutput{}, err
+	}
+
+	var sum uint64
+	for _, piece := range pieces {
```

Comment:
these pieces include the padded size, correct?

Reply:
Yes, that's the `UserBytes(NearestPowerOf2(size))` size